### PR TITLE
Add support for restoring a given namespace

### PR DESCRIPTION
**Pull request by Lujie1996**
## Overview

Description:
This PR adds a new restore mode called PARTIAL_NAMESPACE, which restores all tables in the given namespaces.

Why should this be merged: 
Required by multi-tenancy.

Related issue(s) (if applicable): #<number>
N/A

## Checklist (Definition of Done):

- [x] There are no TODOs left in the code
- [x] [Coding conventions](https://github.com/CorfuDB/CorfuDB/wiki/Corfu-Style-Guidelines) (e.g. for logging, unit tests) have been followed
- [x] Change is covered by automated tests
- [x] Public API has Javadoc

### CODE REVIEW

**Review by chetangudisagar — APPROVED** [review]
LGTM.

File: runtime/src/main/java/org/corfudb/runtime/Restore.java [hunk 1]
```diff
@@ -136,23 +146,31 @@ private void enableCompaction() {
...
             try {
-                Path tableBackupPath = Paths.get(restoreTempDirPath).resolve(tableBackup);
+                Path tableBackupPath = Paths.get(restoreTempDirPath).resolve(tableBackupFileName.toString());
                 restoreTable(tableBackupPath, streamId);
             } catch (IOException e) {
-                log.error("failed to restore table {} from temp file {}", streamId, tableBackup);
+                log.error("failed to restore table {} from temp file {}", streamId, tableBackupFileName);
                 throw e;
             }
             restoredTableCount++;
```
Comment (SravanthiAshokKumar) [inline runtime/src/main/java/org/corfudb/runtime/Restore.java:176]: nit: Can we also print totalNumTables in L179? Might be handy for debugging.
"successfully restored X out of Y tables"

File: runtime/src/main/java/org/corfudb/runtime/Restore.java [hunk 2]
```diff
@@ -136,23 +146,31 @@ private void enableCompaction() {
...
-            if (restoreMode == RestoreMode.PARTIAL_TAGGED && !isTableTagged(tableBackup)) {
-                log.info("skip restoring table {} since it doesn't have requires_backup_support tag", tableBackup);
+        for (TableBackupFileName tableBackupFileName : tableBackupFileNames) {
+            log.info("start restoring table {}", tableBackupFileName);
+            if (restoreMode == RestoreMode.PARTIAL_TAGGED && !isTableTagged(tableBackupFileName)) {
+                log.info("skip restoring table {} since it doesn't have requires_backup_support tag", tableBackupFileName);
+                continue;
+            }
+
+            if (restoreMode == RestoreMode.PARTIAL_NAMESPACE &&
```
Comment (SravanthiAshokKumar) [inline runtime/src/main/java/org/corfudb/runtime/Restore.java:162]: Just for clarification: In restore it looks like we can do either PARTIAL_TAGGED only or PARTIAL_NAMESPACE only. But in backup we can do an intersection of taggedTablesOnly and namespace. Not sure of the use cases but is this intended?

File: runtime/src/main/java/org/corfudb/runtime/Restore.java [hunk 3]
```diff
@@ -205,19 +223,17 @@ private void restoreTable(Path filePath, UUID streamId) throws IOException {
     /**
      * Check if the table has requires_backup_tag. Return true if it's RegistryTable.
      */
-    private boolean isTableTagged(String tableBackup) {
-        // tableBackup name format: uuid.namespace$tableName
-        UUID uuid = UUID.fromString(tableBackup.substring(0, tableBackup.indexOf(".")));
+    private boolean isTableTagged(TableBackupFileName tableBackupFileName) {
```
Comment (SravanthiAshokKumar) [inline runtime/src/main/java/org/corfudb/runtime/Restore.java:226]: nit: For clarity, add a comment to explain this is calculated based off of the restored registryTable

File: runtime/src/main/java/org/corfudb/runtime/Restore.java [hunk 4]
```diff
@@ -16,6 +16,8 @@
 import org.corfudb.runtime.view.TableRegistry;
 import org.corfudb.util.serializer.Serializers;
 
+import javax.annotation.Nonnull;
```
Comment (xnull) [inline runtime/src/main/java/org/corfudb/runtime/Restore.java:19]: Should it be lombok?

**Review by xnull — APPROVED** [review]
looks good, please double check imports

File: runtime/src/main/java/org/corfudb/runtime/Restore.java (hunk 4 above)
Comment (Lujie1996) [reply runtime/src/main/java/org/corfudb/runtime/Restore.java:19]: We have these similar annotations from both lombok and javax in Corfu. Are you suggesting that we should prefer the lombok ones?

File: runtime/src/main/java/org/corfudb/runtime/Restore.java (hunk 1 above)
Comment (Lujie1996) [reply runtime/src/main/java/org/corfudb/runtime/Restore.java:176]: Done

File: runtime/src/main/java/org/corfudb/runtime/Restore.java (hunk 3 above)
Comment (Lujie1996) [reply runtime/src/main/java/org/corfudb/runtime/Restore.java:226]: Done.

File: runtime/src/main/java/org/corfudb/runtime/Restore.java (hunk 2 above)
Comment (Lujie1996) [reply runtime/src/main/java/org/corfudb/runtime/Restore.java:162]: Currently there is no detailed requirements for the namespace related APIs. What you mentioned could be requested in the future and will be implemented upon requests.

**Review by SravanthiAshokKumar — APPROVED** [review]

**Pull request merged**